edgexfoundry/device-mqtt-go
# feat: Enable use of secrets via SecretProvider for MQTT broker credentials

## PR Checklist
Please check if your PR fulfills the following requirements:


## PR Type
What kind of change does this PR introduce?




## What is the current behavior?


MQTT broker credentials are in plain text configuration. No way to get them from SecretStore (aka Vault)

Issue Number: #159


## What is the new behavior?
Via use of SpecretProvider, MQTT broker credentials can be pulled from a SecretStore. The SecretStore is Vault when running in secure mode or InsecureSecrets configuration section when running non-secure mode.

This PR is dependent on the following PRs:

https://github.com/edgexfoundry/go-mod-bootstrap/pull/141
https://github.com/edgexfoundry/device-sdk-go/pull/707

## Does this PR introduce a breaking change?



## Are there any new imports or modules? If so, what are they used for and why?

no

## Are there any specific instructions or things that should be known prior to reviewing?

Do the following steps to test this PR:
1. Cone developer-scripts 
2. cd to releases/nightly-build/compose-files
3. Create password file and config file for MQTT
   - run `mkdir mqtt-config`
   - run `echo 'edgex:$6$OV0KYbPEN4xbWLt9$MkEhwMRMZ8tTYvrlKHnUmZDJtzJGN1RcKNnoM1jNm7zzSgwQo9M0aAAB/8oTqCSQyVy1a42L7jO9xOsuNC9uhg==' > mqtt-config/passwords`
   - run `echo 'allow_anonymous false' > mqtt-config/mosquitto.conf`
   - run `echo 'password_file /mosquitto/config/passwords' >> mqtt-config/mosquitto.conf`
3. Add MQTT Broker 
   - Add the following to nightly-build secure compose file (docker-compose-nexus.yml)
```
  mqtt-broker:
    container_name: edgex-mqtt-broker
    hostname: edgex-mqtt-broker
    image: eclipse-mosquitto:1.6.3
    networks:
      edgex-network: { }
    ports:
      - 1883:1883/tcp
    volumes:
    - ./mqtt-config:/mosquitto/config
```
4. Add `ADD_SECRETSTORE_TOKENS: "device-mqtt"` to Vault Work in compose file.
5. Run the compose file
   - `make run` 
6. MQTT Broker is now running with user/password auth
  - User is `edge` and Password is `password`
  - Use a tool like MQTT.fx to connect to verify the user and password are set correctly and are required.
6. clone the branch for this PR
7. Build Device MQTT
8. cd cmd
9. run `sudo WRITABLE_LOGLEVEL=DEBUG EDGEX_SECURITY_SECRET_STORE=true ./device-mqtt -c ./res/example`
  - Expect to see DEBUG message retrying `Using Secrets URL of  `http://localhost:8200/v1/secret/edgex/device-mqtt/credentials`
  - Expect to see ERROR message like "Unable to retrieve MQTT credentials from SecretProvider at path 'credentials': Error found on handling secrets from underlying data-store: Received a '404' response from the secret store. Retrying for 9.9559743s"
10 Store MQTT credentials
   - You have 60secs to store the MQTT credential via /api/v2/secret endpoint before service fails.
     - This can be increased in config via `Driver.CredentialsRetryTime` setting
   - While DEBUG messages are being logged send the following via REST POST to `http://localhost:49982/api/v2/secret`
```
{
  "path" : "credentials",
  "secretData" : [
    {
      "key" : "username",
      "value" : "edgex"
    },
	{
      "key" : "password",
      "value" : "password"
    }
  ]
}
```
     - Expected response it:
```
{
    "apiVersion": "v2",
    "requestId": "",
    "message": "",
    "statusCode": 201
}
```
   - The following messages will then be logged:
```
 "[Response listener] Start command response listening."
 "[Incoming listener] Start incoming data listening. "
```
  - Device Service is ready to accept async data from a device
10. From a tool like MQTT.fx publish the following to the `DataTopic` topic
```
{
  "name" : "MQTT-test-device",
  "cmd" : "message",
  "message" : "Hellow World"
}
```
  - `"SendEvent: Pushed event to core data"` should be logged.
11.  Use core-data REST API to confirm reading received by sending GET to `localhost:48080/api//v1/reading/name/message/0`

## Other information
